ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -553,140 +547,176 @@ let definition_query (state : State.t) uri position merlin_request =
  fun rpc req ->
   let state = Server.state rpc in
   let store = state.store in
+  let now res = Fiber.return (Reply.now (Ok res), state) in
+  let later f req =
+    Fiber.return
+      ( Reply.later (fun k ->
+            let open Fiber.O in
+            let* resp = f state req in
+            k resp)
+      , state )
+  in
+  let not_supported = Fiber.return (Reply.now not_supported, state) in
+  let error e = Fiber.return (Reply.now (Error e), state) in
   match req with
-  | Client_request.Initialize ip -> on_initialize rpc ip
-  | Client_request.Shutdown -> Fiber.return @@ Ok ((), state)
+  | Client_request.Initialize ip ->
+    let res, state = on_initialize rpc ip in
+    Fiber.return (Reply.now (Ok res), state)
+  | Client_request.Shutdown -> now ()
   | Client_request.DebugTextDocumentGet { textDocument = { uri }; position = _ }
-    ->
-    Fiber.return
-      (let uri = Uri.t_of_yojson (`String uri) in
-       match Document_store.get_opt store uri with
-       | None -> Ok (None, state)
-       | Some doc -> Ok (Some (Msource.text (Document.source doc)), state))
-  | Client_request.DebugEcho params -> Fiber.return @@ Ok (params, state)
-  | Client_request.TextDocumentColor _ -> Fiber.return @@ Ok ([], state)
-  | Client_request.TextDocumentColorPresentation _ ->
-    Fiber.return @@ Ok ([], state)
-  | Client_request.TextDocumentHover req -> hover state req
-  | Client_request.TextDocumentReferences req -> references state req
-  | Client_request.TextDocumentCodeLensResolve codeLens ->
-    Fiber.return @@ Ok (codeLens, state)
-  | Client_request.TextDocumentCodeLens req -> text_document_lens state req
-  | Client_request.TextDocumentHighlight { textDocument = { uri }; position } ->
-    let open Fiber.Result.O in
+    -> (
     let uri = Uri.t_of_yojson (`String uri) in
-    let* doc = Fiber.return (Document_store.get store uri) in
-    let command =
-      Query_protocol.Occurrences (`Ident_at (Position.logical position))
-    in
-    let open Fiber.O in
-    let+ locs = Document.dispatch_exn doc command in
-    let lsp_locs =
-      List.map locs ~f:(fun loc ->
-          let range = Range.of_loc loc in
-          (* using the default kind as we are lacking info to make a difference
-             between assignment and usage. *)
-          DocumentHighlight.create ~range ~kind:DocumentHighlightKind.Text ())
-    in
-    Ok (Some lsp_locs, state)
-  | Client_request.WorkspaceSymbol _ -> Fiber.return @@ Ok (None, state)
+    match Document_store.get_opt store uri with
+    | None -> now None
+    | Some doc -> now (Some (Msource.text (Document.source doc))) )
+  | Client_request.DebugEcho params -> now params
+  | Client_request.TextDocumentColor _ -> now []
+  | Client_request.TextDocumentColorPresentation _ -> now []
+  | Client_request.TextDocumentHover req -> later hover req
+  | Client_request.TextDocumentReferences req -> later references req
+  | Client_request.TextDocumentCodeLensResolve codeLens -> now codeLens
```

Comment:
What's the reason for not running codeLens concurrently?

Reply:
This is `TextDocumentCodeLensResolve` rather code lens. We don't do anything interesting on this request anyway (yet)